pkg/controller/istiocontrolplane/listeners.go

```diff
@@ -243,6 +246,77 @@ func (c *IstioDefaultChartCustomizer) waitForDeployment(object runtime.Object) {
 	}
 }
 
+// CitadelChartCustomizer is a ChartCustomizer for the citadel chart
+type CitadelChartCustomizer struct {
+	*IstioDefaultChartCustomizer
+}
+
+var _ helmreconciler.ChartCustomizer = &CitadelChartCustomizer{}
+
+// NewCitadelChartCustomizer creates a new CitadelChartCustomizer
+func NewCitadelChartCustomizer(chartName, chartAnnotationKey string) *CitadelChartCustomizer {
+	return &CitadelChartCustomizer{
+		IstioDefaultChartCustomizer: NewIstioDefaultChartCustomizer(chartName, chartAnnotationKey),
+	}
+}
+
+// ResourceDeleted invokes the default ResourceDeleted behavior for all resources and delete the "istio-security"
+// configmap and related secrets generated by citadel.
+func (c *CitadelChartCustomizer) ResourceDeleted(obj runtime.Object) error {
+	var err error
+	if err = c.IstioDefaultChartCustomizer.ResourceDeleted(obj); err != nil {
+		return err
+	}
+	switch obj.GetObjectKind().GroupVersionKind().Kind {
+	case "Deployment":
+		if err = c.cleanCitadelResources(obj); err != nil {
+			log.Errorf("error cleaning the citadel resources: %s", err)
+			return err
+		}
+	}
+	return err
+}
+
+func (c *CitadelChartCustomizer) cleanCitadelResources(obj runtime.Object) error {
+	cleanCitadelResourcesMaxRetries := 3
+	objAccessor, err := meta.Accessor(obj)
+	if err != nil {
+		return fmt.Errorf("get error to get namespace of %s: %s", obj.GetObjectKind().GroupVersionKind(), err)
+	}
+
+	client := c.Reconciler.GetClient()
+	cmNamespace := objAccessor.GetNamespace()
+	cmName := "istio-security"
+	prunedSecretType := "istio.io/key-and-cert"
+
+	cm := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cmName,
+			Namespace: cmNamespace,
+		},
+	}
+
+	secretList := &corev1.SecretList{}
+	if err := client.List(context.TODO(), secretList); err != nil {
+		return fmt.Errorf("error getting secret list: %s", err)
+	}
+
+	for retryCount := 1; retryCount <= cleanCitadelResourcesMaxRetries; retryCount++ {
+		if err := client.Delete(context.TODO(), cm); err != nil && retryCount == cleanCitadelResourcesMaxRetries {
+			return fmt.Errorf("error deleting ConfigMap %s/%s: %s", cmNamespace, cmName, err)
+		}
+
+		for _, secret := range secretList.Items {
```

Comment:
today, I'd you delete istio your sidecars will still function properly. If you delete the secrets I don't think they will

Reply:
I assume this is only in mTLS strict mode perhaps?  I just did a test, when i wipe out my istio-systems ns, my bookinfo still function okay, with sidecars.

When all istio CRDs are removed in our case when the removal of icp cr, i don't see how you enforce istio mTLS strict mode...

Reply:
https://github.com/istio/istio/issues/18031#issuecomment-543502295
does it mean that we only need to delete the secrets during upgrading istio? For istio uninstallation, we should keep these secrets?

Reply:
Do not delete them during upgrade or you will get a cluster wide traffic loss. Once the certs are deleted envoy will hot restart without the certs and start failing mtls calls.

Maybe we don't care about this in deletion, but definitely in upgrade

Reply:
Agree on not deleting during upgrade.

We should delete the secrets when removal of the icp CR when users have decided they don't want to use istio and we return a cleanup for them.

Reply:
Got it, I need to make sure the `ResourceDeleted` will not invoked during upgrade.

Reply:
and keep in mind if we install a new operator version (for z stream updates) - nothing should be deleted - instead the existing resources should just be reconciled.